File: frontend/src/app/services/recipe.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams, HttpResponse} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../global/globals';
import {RecipeDetails} from '../dtos/recipe-details';
import {RecipeSearch} from '../dtos/recipe-search';
import { RecipeListElement } from '../dtos/recipe-list-element';

@Injectable({
  providedIn: 'root'
})
export class RecipeService {

  private recipeBaseUri: string = this.globals.backendUri + '/recipe';

  constructor(
    private httpClient: HttpClient,
    private globals: Globals
    ) { }

  /**
   * Loads the recipes at the specified page from the backend
   */
  getRecipes(page: number = 0, size: number = 20): Observable<HttpResponse<RecipeListElement[]>> {
    let pageParams = new HttpParams();
    pageParams = pageParams.append('size', size);
    pageParams = pageParams.append('page', page);

    return this.httpClient.get<RecipeListElement[]>(this.recipeBaseUri, {params: pageParams, observe: 'response'});
  }

  /**
   * Search for recipes
   *
   * @param search The recipes to search for
   * @returns Returns a list of recipes matching the search parameters
   */
  getSearchResults(page: number = 0, size: number = 20, search: RecipeSearch): Observable<HttpResponse<RecipeListElement[]>> {
    let pageParams = new HttpParams();
    pageParams = pageParams.append('size', size);
    pageParams = pageParams.append('page', page);

    if(search.difficulty !== null){
      pageParams = pageParams.append('difficulty', search.difficulty);
    }

    if(search.name !== null){
      pageParams = pageParams.append('name', search.name);
    }

    return this.httpClient.get<RecipeListElement[]>(this.recipeBaseUri + '/search', {params: pageParams, observe: 'response'});
  }

  /**
   * Loads specific recipe from the backend
   *
   * @param id of recipe to load
   */
  getRecipeById(id: number): Observable<RecipeDetails> {
    return this.httpClient.get<RecipeDetails>(this.recipeBaseUri + '/' + id);
  }

  /**
   * Persists recipe to the backend
   *
   * @param recipe to persist
   */
  createRecipe(recipe: RecipeDetails): Observable<void> {
    return this.httpClient.post<void>(this.recipeBaseUri, recipe);
  }

  /**
   * Updates a recipe in the database
   *
   * @param recipe The new data of the wine
   * @param id The id of the wine to update
   * @returns Returns the updated wine
   */
  updateRecipe(recipe: RecipeDetails, id: number): Observable<RecipeDetails> {
    return this.httpClient.post<RecipeDetails>(this.recipeBaseUri + '/' + id, recipe);
  }

  /**
   * Delete the specified wine
   *
   * @param id The id of the wine to delete
   */
  deleteRecipe(id: number): Observable<void> {
    return this.httpClient.delete<void>(this.recipeBaseUri + '/' + id);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/RecipeService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.RecipeCreateDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.RecipeDetailsDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.RecipeSearchDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Recipe;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ObjectAlreadyExistsException;
import at.ac.tuwien.sepm.groupphase.backend.exception.UserPermissionException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ValidationException;
import org.springframework.data.domain.Pageable;


import java.util.List;

/**
 * This service contains everything related to the management (CRUD) of recipes.
 */
public interface RecipeService {

    /**
     * Fetch a recipe by its id.
     *
     * @param id The id of the wine to look for.
     * @return Returns the recipe
     * @throws NotFoundException If the recipe could not be found
     */
    Recipe getRecipeById(Long id) throws NotFoundException;

    /**
     * Fetch all recipes for the specified page.
     *
     * @param pageable The pages to get
     * @return Returns the recipes
     */
    List<Recipe> getRecipes(Pageable pageable);

    /**
     * Create a new recipe with the calling user as the owner.
     *
     * @param wine     The wine to create
     * @param username The owner of the recipe
     * @throws ValidationException          Thrown if the validation of the createDto failed
     * @throws ObjectAlreadyExistsException Thrown if a similar wine already exists.
     */
    void createRecipe(RecipeCreateDto wine, String username) throws ValidationException, ObjectAlreadyExistsException;

    /**
     * Updates the recipe based on the specified information.
     *
     * @param updated  The update recipe details
     * @param username The username of the user that called this action
     * @return Returns the updated recipe
     * @throws NotFoundException            Thrown if the dto to update is not found
     * @throws UserPermissionException      Thrown if the user that executed this query is not the owner or an admin
     * @throws ValidationException          Thrown if the Details dto failed validation
     * @throws ObjectAlreadyExistsException Thrown if a similar recipe already exists.
     */
    Recipe updateRecipe(RecipeDetailsDto updated, String username) throws NotFoundException,
        UserPermissionException,
        ValidationException,
        ObjectAlreadyExistsException;

    /**
     * Deletes the recipe.
     *
     * @param id       The id of the recipe
     * @param username The user who called this query
     * @throws NotFoundException       Thrown if the requested recipe could not be found
     * @throws UserPermissionException Thrown if the user is not the owner or an admin
     */
    void delete(Long id, String username) throws NotFoundException, UserPermissionException;

    /**
     * Count all recipes in the table.
     *
     * @return Returns the number of wines in the database
     */
    long countAll();


    /**
     * Count all recipes matching the search parameters.
     *
     * @param searchDto The search parameters to look for
     * @return Returns the number of wines matching the search query.
     */
    long countFindBy(RecipeSearchDto searchDto) throws ValidationException;

    /**
     * Fetch all recipes for the specified page and search.
     *
     * @param recipeSearchDto The search parameters to look for
     * @param pageable        The pages to get
     * @return Returns the wines
     */
    List<Recipe> findRecipes(RecipeSearchDto recipeSearchDto, Pageable pageable) throws ValidationException;
}
